crates/wysiwyg/src/format_nodes.rs

```diff
@@ -0,0 +1,405 @@
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_start() {
+        let model = cm("hell{o <b>w}|orld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_end() {
+        let model = cm("<b>hell{o </b>wor}|ld");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_extending_middle() {
+        let model = cm("<b>h{el</b>lo <b>wor}|ld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Extend);
+    }
+
+    #[test]
+    fn selection_type_remove() {
+        let model = cm("<b>hel{lo </b><b>wor}|ld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Remove);
+    }
+
+    #[test]
+    fn selection_type_remove_on_start_edge() {
+        let model = cm("{<b>hello </b><b>wor}|ld</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Remove);
+    }
+
+    #[test]
+    fn selection_type_remove_on_ending_edge() {
+        let model = cm("<b>hel{lo </b><b>world}|</b>");
+        let range = find_multiple_range(&model);
+        let selection_type = model.check_format_selection_type(
+            &range.locations,
+            &InlineFormatType::Bold,
+        );
+        assert_eq!(selection_type, FormatSelectionType::Remove);
+    }
+
+    #[test]
+    fn formatting_several_nodes_works_with_different_format() {
+        let mut model = cm("{hello <i>wor}|ld</i>");
+        model.format(InlineFormatType::Bold);
+        assert_eq!(
+            model.state.dom.to_string(),
+            "<strong>hello </strong><i><strong>wor</strong>ld</i>"
+        );
+    }
+
+    #[test]
+    fn formatting_several_nodes_works_with_same_format() {
+        let mut model = cm("{hello <b>wor}|ld</b>");
+        model.format(InlineFormatType::Bold);
+        assert_eq!(model.state.dom.to_string(), "<strong>hello world</strong>");
```

Comment:
I am really happy it works like this (transforming b to strong in this case, but not when not needed). Awesome.